File: go/cli/compile.go
package cli

import (
	"flag"
	"io"

	compiler "github.com/ball-lang/ball/go/compiler"
)

// cmdCompile implements `ball compile <program.ball.json> [-o out.go]`: load the
// program and compile it to runnable Go source via go/compiler, writing to -o or
// stdout.
//
// The compiler fails loud (issue #55): an unsupported expression shape or base
// function is returned as an error, surfaced here as a parseErr (exit 2), never
// silent bad code. The emitted source is gofmt'd for readability; a formatting
// failure on otherwise-valid output is non-fatal (the raw source is still
// written) — only a real compile error stops the command.
func cmdCompile(args []string, w io.Writer) *cliError {
	const usage = "ball compile <program.ball.json> [-o out.go]"
	fs := flag.NewFlagSet("compile", flag.ContinueOnError)
	out := fs.String("o", "", "write the generated Go source here instead of stdout")
	positionals, cerr := parseCommand(fs, "compile", usage, args)
	if cerr != nil {
		return cerr
	}
	if len(positionals) != 1 {
		return parseErr("compile: expected exactly one program path (usage: %s)", usage)
	}

	prog, cerr := loadProgram(positionals[0])
	if cerr != nil {
		return cerr
	}

	src, compileErr := compiler.Compile(prog)
	if formatted, ferr := compiler.Format(src); ferr == nil {
		src = formatted
	}
	if compileErr != nil {
		return parseErr("compile: %v", compileErr)
	}
	return writeOut(w, *out, []byte(src))
}

File: python/cli/ball_cli/commands/compile.py
"""``ball compile <program.ball.json> [-o out.py]`` — Ball -> Python source.

Loads the program and compiles it to runnable Python via ``python/compiler``,
writing to ``-o`` or stdout. The compiler fails loud (issue #55): an unsupported
expression shape or base function raises ``CompileError``, surfaced here as an
invalid-program error (exit 2), never silently-wrong code.
"""

from __future__ import annotations

from typing import TextIO

from ..argparse_util import StreamParser
from ..errors import parse_error
from ..loader import load_program_dict
from ..output import write_output


def command(args: list[str], stdout: TextIO, stderr: TextIO) -> int:
    parser = StreamParser(
        prog="ball compile",
        description="Compile a Ball program (.ball.json) to Python source.",
        out=stdout,
        err=stderr,
    )
    parser.add_argument("input", metavar="<program.ball.json>", help="path to a .ball.json program")
    parser.add_argument(
        "-o", "--output", metavar="<out.py>",
        help="write the generated Python source here instead of stdout",
    )
    ns = parser.parse_args(args)

    program = load_program_dict(ns.input)

    from ball_compiler import CompileError, compile_program

    try:
        source = compile_program(program)
    except CompileError as ex:
        raise parse_error(f"compile: {ex}") from ex

    write_output(ns.output, source, stdout)
    return 0
